File: nym-vpn-android/app/src/main/java/net/nymtech/nymvpn/auth/LabAuthClient.kt
package net.nymtech.nymvpn.auth

import android.util.Log
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import net.nymtech.nymvpn.BuildConfig
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL

/** Lab-only username/password auth against nym-mock-api. */
object LabAuthClient {
	private const val TAG = "lab-auth"
	private val json = Json { ignoreUnknownKeys = true }

	private val baseUrl: String
		get() = BuildConfig.LAB_AUTH_BASE_URL.trimEnd('/')

	@Serializable
	private data class CredRequest(
		val username: String,
		val password: String,
	)

	@Serializable
	private data class LoginResponse(
		val username: String,
		val mnemonic: String,
	)

	@Serializable
	private data class RegisterResponse(
		val username: String,
		val message: String? = null,
	)

	@Serializable
	private data class ErrorResponse(
		val error: String? = null,
	)

	sealed class LabAuthException(message: String) : IOException(message) {
		class BadRequest(msg: String) : LabAuthException(msg)
		class Conflict(msg: String) : LabAuthException(msg)
		class Unauthorized(msg: String) : LabAuthException(msg)
		class Server(msg: String) : LabAuthException(msg)
	}

	suspend fun register(username: String, password: String) {
		post("register", CredRequest(username, password), RegisterResponse.serializer(), 201)
	}

	suspend fun login(username: String, password: String): String {
		val resp = post("login", CredRequest(username, password), LoginResponse.serializer(), 200)
		return resp.mnemonic
	}

	private suspend fun <T> post(path: String, body: CredRequest, deserializer: kotlinx.serialization.KSerializer<T>, okCode: Int): T {
		return kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.IO) {
			val url = URL("$baseUrl/$path")
			val conn = (url.openConnection() as HttpURLConnection).apply {
				requestMethod = "POST"
				connectTimeout = 15_000
				readTimeout = 15_000
				doOutput = true
				setRequestProperty("Content-Type", "application/json; charset=utf-8")
			}
			try {
				conn.outputStream.bufferedWriter().use { it.write(json.encodeToString(CredRequest.serializer(), body)) }
				val code = conn.responseCode
				val stream = if (code in 200..299) conn.inputStream else conn.errorStream
				val text = stream?.bufferedReader()?.readText().orEmpty()
				Log.d(TAG, "POST $path -> $code")
				if (code == okCode) {
					return@withContext json.decodeFromString(deserializer, text)
				}
				val err = runCatching { json.decodeFromString(ErrorResponse.serializer(), text).error }.getOrNull()
				val msg = err ?: "HTTP $code"
				throw when (code) {
					400 -> LabAuthException.BadRequest(msg)
					401, 403 -> LabAuthException.Unauthorized(msg)
					409 -> LabAuthException.Conflict(msg)
					else -> LabAuthException.Server(msg)
				}
			} finally {
				conn.disconnect()
			}
		}
	}
}

File: nym-vpn-apple/Services/Sources/Services/ConfigurationManager/LabAuthClient.swift
import Foundation
import Constants

/// Lab-only username/password auth against nym-mock-api (Android `LabAuthClient` parity).
public enum LabAuthClient {
    public enum LabAuthError: LocalizedError {
        case disabled
        case badRequest(String)
        case conflict(String)
        case unauthorized(String)
        case server(String)
        case invalidResponse

        public var errorDescription: String? {
            switch self {
            case .disabled:
                return "Lab auth is only available in lab mock builds"
            case .badRequest(let message),
                 .conflict(let message),
                 .unauthorized(let message),
                 .server(let message):
                return message
            case .invalidResponse:
                return "Invalid lab auth response"
            }
        }
    }

    private struct CredRequest: Encodable {
        let username: String
        let password: String
    }

    private struct LoginResponse: Decodable {
        let username: String
        let mnemonic: String
    }

    private struct ErrorResponse: Decodable {
        let error: String?
    }

    public static func register(username: String, password: String) async throws {
        guard LabMock.isEnabled else { throw LabAuthError.disabled }
        let trimmed = username.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, !password.isEmpty else {
            throw LabAuthError.badRequest("username and password required")
        }
        guard password.count >= 8 else {
            throw LabAuthError.badRequest("Password must be at least 8 characters")
        }
        try await post(path: "register", body: CredRequest(username: trimmed, password: password), okStatus: 201)
    }

    public static func login(username: String, password: String) async throws -> String {
        guard LabMock.isEnabled else { throw LabAuthError.disabled }
        let trimmed = username.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, !password.isEmpty else {
            throw LabAuthError.badRequest("username and password required")
        }
        let data = try await post(
            path: "login",
            body: CredRequest(username: trimmed, password: password),
            okStatus: 200
        )
        let response = try JSONDecoder().decode(LoginResponse.self, from: data)
        let mnemonic = response.mnemonic.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !mnemonic.isEmpty else { throw LabAuthError.invalidResponse }
        return mnemonic
    }

    @discardableResult
    private static func post(path: String, body: CredRequest, okStatus: Int) async throws -> Data {
        let base = LabMock.authBaseURL
        guard let url = URL(string: "\(base)/\(path)") else {
            throw LabAuthError.server("Invalid lab auth URL")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 15
        request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await URLSession.shared.data(for: request)
        let code = (response as? HTTPURLResponse)?.statusCode ?? -1
        if code == okStatus {
            return data
        }

        let message = (try? JSONDecoder().decode(ErrorResponse.self, from: data))?.error
            ?? "HTTP \(code)"
        switch code {
        case 400:
            throw LabAuthError.badRequest(message)
        case 401, 403:
            throw LabAuthError.unauthorized(message)
        case 409:
            throw LabAuthError.conflict(message)
        default:
            throw LabAuthError.server(message)
        }
    }
}
